src/AppInstallerRepositoryCore/RepositorySource.cpp

```diff
@@ -446,6 +396,158 @@ namespace AppInstaller::Repository
 
             return false;
         }
+
+        // Struct containing internal implementation of source list
+        // This contains all sources including tombstoned sources
+        struct SourceListInternal
+        {
+            SourceListInternal();
+
+            // Get a list of current sources references which can be used to update the contents in place.
+            // e.g. update the LastTimeUpdated value of sources.
+            std::vector<std::reference_wrapper<SourceDetailsInternal>> GetCurrentSourceRefs();
+
+            // Current source means source that's not in tombstone
+            bool ContainsCurrentSource(std::string_view name);
+            SourceDetailsInternal& GetCurrentSource(std::string_view name);
+
+            // Source includes ones in tombstone
+            bool ContainsSource(std::string_view name);
+            SourceDetailsInternal& GetSource(std::string_view name);
+
+            // Add/remove a current source
+            void AddSource(const SourceDetailsInternal& source);
+            void RemoveSource(const SourceDetailsInternal& source);
+
+            // Save source metadata. Currently only LastTimeUpdated is used.
+            void SaveMetadata();
+
+        private:
+            std::vector<SourceDetailsInternal> m_sourceList;
+
+            // calls std::find_if and return the iterator.
+            auto FindSource(std::string_view name, bool includeTombstone = false);
+        };
+
+        SourceListInternal::SourceListInternal()
+        {
+            for (SourceOrigin origin : { SourceOrigin::User, SourceOrigin::Default })
+            {
+                auto forOrigin = GetSourcesByOrigin(origin);
+
+                for (auto&& source : forOrigin)
+                {
+                    if (!ContainsSource(source.Name))

```

Comment:
>ContainsSource [](start = 25, length = 14)

It would be more efficient here if `GetSource` could return an invalid value.  That way you don't need to search twice (applies to `LastUpdateTime` below even more so).